ws/api/endpoint_test.go

```
@@ -0,0 +1,187 @@
+			chanID:   "",
+			subtopic: "",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusBadRequest,
+			msg:      []byte{},
+		},
+		{
+			name:     "connect with empty thingKey",
+			chanID:   id,
+			subtopic: "",
+			header:   true,
+			thingKey: "",
+			status:   http.StatusForbidden,
+			msg:      []byte{},
+		},
+		{
+			name:     "connect and send message with thingKey as query parameter",
+			chanID:   id,
+			subtopic: "",
+			header:   false,
+			thingKey: thingKey,
+			status:   http.StatusSwitchingProtocols,
+			msg:      msg,
+		},
+		{
+			name:     "connect and send message that cannot be published",
+			chanID:   id,
+			subtopic: "",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusSwitchingProtocols,
+			msg:      []byte{},
+		},
+		{
+			name:     "connect and send message to subtopic",
+			chanID:   id,
+			subtopic: "subtopic",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusSwitchingProtocols,
+			msg:      msg,
+		},
+		{
+			name:     "connect and send message to subtopic with invalid name",
+			chanID:   id,
+			subtopic: "sub/a*b/topic",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusBadRequest,
+			msg:      msg,
+		},
+		{
+			name:     "connect and send message to nested subtopic",
+			chanID:   id,
+			subtopic: "subtopic/nested",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusSwitchingProtocols,
+			msg:      msg,
+		},
+		{
+			name:     "connect and send message to all subtopics",
+			chanID:   id,
+			subtopic: ">",
+			header:   true,
+			thingKey: thingKey,
+			status:   http.StatusSwitchingProtocols,
+			msg:      msg,
+		},
+	}
+
+	for _, tt := range cases {
+		conn, res, err := handshake(ts.URL, tt.chanID, tt.subtopic, tt.thingKey, tt.header)
+		assert.Equal(t, tt.status, res.StatusCode, fmt.Sprintf("expected status code '%d' got '%d'\n", tt.status, res.StatusCode))
+		if err != nil {
+			return
+		}
```

This must be changed - it makes tests fail silently. You can add an expected error to the test case and do something like:
```go
assert.Equal(t, err, tc.err, fmt.Sprintf(....))
if tc.err != nil {
		err = conn.WriteMessage(websocket.TextMessage, tt.msg)
		assert....
}
```